CORDA-3809 Expose client side unique RPC ID for flow starts

kyriathar opened the pull request:
Introducing a new flow start method (`startFlowDynamicWithClientId`) passing in a `clientId`. 

If a new flow starts with this method, then a < `clientId` to flow > pair is kept on node side, even after the flow's lifetime. If `startFlowDynamicWithClientId` is called again with the same `clientId` then the node identifies that this `clientId` refers to an existing < `clientId` to flow > pair, and returns back to the rpc client a `FlowStateMachineHandle` future, created out of that pair.

Introducing an ackowledgement method (`removeClientId`). Calling this method removes the < `clientId` to flow > pair on node side and frees the resources.

kyriathar-r3 committed:
First attempt to pass client unique id inside the statemachine

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -264,6 +265,9 @@ interface CordaRPCOps : RPCOps {
     fun <T> startFlowDynamic(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T>
 
+    @RPCReturnsObservables
+    fun <T> startFlowDynamicWithClientId(logicType: Class<out FlowLogic<T>>, clientId: UUID, vararg args: Any?): FlowHandle<T>
```
I wonder if we should create a sub-interface of `FlowHandle` called `FlowHandleAndClientId` that also exposes the supplied client Id ?  Might be useful if you have a collection of returned values to know what the client ID for each one is.

rick-r3 replied:
As discussed, if we put clientID first the rest of the signature can mirror the existing function.

rick-r3 replied:
Hmmm.... I thought clientID was going to be a String ?  That would be better.  What did we do in the schema?

kyriathar-r3 committed:
Move clientId at the beginning of methods

kyriathar-r3 committed:
Change clientId type from UUID to String

kyriathar-r3 committed:
Make getFlowParameters expect 2 or 3 arguments

kyriathar-r3 committed:
Minor arguments' reordering to match parameters order

kyriathar-r3 committed:
Revert "Make getFlowParameters expect 2 or 3 arguments"

This reverts commit 25608f2cb677106742b403dca2a67b3a02ecd3aa.

kyriathar-r3 committed:
Revert "Change clientId type from UUID to String"

This reverts commit a8079083e2200f0a9616fc580061ed15e57969ea.

kyriathar-r3 committed:
Revert "Move clientId at the beginning of methods"

This reverts commit a9254ba760316f789f29727da3ca56a719b4bc60.

kyriathar-r3 committed:
Revert "First attempt to pass client unique id inside the statemachine"

This reverts commit 9b2703c106e1d29f5e5834bba750b1f42d0f1ae5.

kyriathar-r3 committed:
Add CordaRPCOps start flow dynamic method with clientId, and pass clientId to InvocationContext

kyriathar-r3 committed:
Revert "Revert "Make getFlowParameters expect 2 or 3 arguments""

This reverts commit ee8099f34952473bfcf76b1756ac3fa9b043400b.

kyriathar-r3 committed:
Add client id as a shortcut var to FlowStateMachine

kyriathar-r3 committed:
Add client it to the logging context of the thread running a fiber

kyriathar-r3 committed:
Rename var

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/logging/ContextualLoggingUtils.kt:
```diff
@@ -39,4 +39,11 @@ internal fun InvocationOrigin.pushToLoggingContext(prefix: String = "") {
     }
+}
+
+internal fun pushClientIdToLoggingContext(clientUUID: String) {
```
I think this should just be called from `pushToLoggingContext` at the top of this file if `clientID != null`, and not where you currently call it.

rick-r3 replied:
But even if you retain the call site above, still also call it from `pushToLoggingContext` and remove the individual call below.

kyriathar replied:
Moved it inside `pushToLoggingContext`.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -267,6 +270,7 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
         MDC.put("flow-id", id.uuid.toString())
         MDC.put("fiber-id", this.getId().toString())
         MDC.put("thread-id", Thread.currentThread().id.toString())
+        clientUUID?.let { pushClientIdToLoggingContext(it) }
```
No need for this.  Rely on `pushToLoggingContext()` above.

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -248,16 +249,22 @@ internal class CordaRPCOpsImpl(
 
-    private fun <T> startFlow(logicType: Class<out FlowLogic<T>>, args: Array<out Any?>): FlowStateMachine<T> {
+    override fun <T> startFlowDynamicWithClientId(clientUUID: String, logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T> {
+        pushClientIdToLoggingContext(clientUUID)
```
I mention elsewhere to get rid of this, but actually does it server a purpose?  e.g. is something logged that is useful?

kyriathar replied:
Indeed it does not log anything. The logging starts when fiber/ flows start on their thread/ threads. Therefore, i m removing it.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -547,7 +548,11 @@ class DBCheckpointStorage(
         return if (arguments.isNotEmpty()) {
-            uncheckedCast<Any?, Array<Any?>>(arguments[1]).toList()
+            when {
+                arguments.size == 2 -> uncheckedCast<Any?, Array<Any?>>(arguments[1]).toList()
```
You should also change line 435 to set `userSuppliedIdentifier`

kyriathar replied:
Yes this is to be added next, along with a test that asserts that is set.

rick-r3 reviewed (commented):
Looking good. See comments

kyriathar-r3 committed:
Remove adding client id to rpc thread context and move client id inside pushToLoggingContext

kyriathar-r3 committed:
Persist client id into node_flow_metadata

rick-r3 commented on core/src/main/kotlin/net/corda/core/context/InvocationContext.kt:
```diff
@@ -49,8 +50,9 @@ data class InvocationContext(
             actor: Actor? = null,
             externalTrace: Trace? = null,
             impersonatedActor: Actor? = null,
+            clientUUID: String? = null,
```
Ah, this should just be `clientID` i.e. not `UUID` !

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -248,16 +248,21 @@ internal class CordaRPCOpsImpl(
 
-    private fun <T> startFlow(logicType: Class<out FlowLogic<T>>, args: Array<out Any?>): FlowStateMachine<T> {
+    override fun <T> startFlowDynamicWithClientId(clientUUID: String, logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T> {
+        val stateMachine = startFlow(logicType, context().withClientId(clientUUID), args)
```
Can we require that the length of the client ID is less than database column size ?

kyriathar replied:
Done -`clientID` is now checked at `FlowStarterImpl`-.

kyriathar-r3 committed:
Rename clientUUID to clientID

kyriathar-r3 committed:
Remove unneeded test; there was already an existing one to assert persisting of client id into into node_flow_metadata

kyriathar-r3 committed:
Adjust test to assert persisting of client id into into node_flow_metadata

kyriathar-r3 committed:
Introduce a thin FlowStateMachine interface consisting of only flowId and result future.

Introduce a new interface extending the above interface and adding the client id.

Make FlowStarterImpl guard that client id length is less or equal than db column length.

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/FlowStateMachine.kt:
```diff
@@ -11,10 +11,15 @@ import net.corda.core.node.ServiceHub
 import net.corda.core.serialization.SerializedBytes
 import org.slf4j.Logger
 
+interface FlowStateMachineReturnable<FLOWRETURN> {
```
I think maybe `FlowStateMachineHandle` would be better.

I see that the "tracked" start flow methods somewhat ruin the abstraction and cause you to need to cast back to `FlowStateMachine`.   How about we put `val logic:  FlowLogic<FLOWRETURN>?` on here and fix up any places that refer to `logic` to require it is not `null` ?  Then we shouldn't need any casting (certainly much less).

I also don't think we can usefully make use of `FlowStateMachineClientIdResult` independently, so I would put the `val clientID: String?` on here too and get rid of `FlowStateMachineClientIdResult`.

If you still need a `as FlowStateMachine`, let's discuss.  I think the only potential candidate might be

`fun <T> StartedNodeServices.startFlow(logic: FlowLogic<T>): FlowStateMachine<T> = startFlow(logic, newContext()).getOrThrow() as FlowStateMachine
`

but I reckon we can make that return `FlowStateMachineHandle` too unless there's a specific test that uses something other than the result.

kyriathar replied:
Done.

rick-r3 reviewed (commented):
Took me a while to figure out why it didn't go as well as I expected.

kyriathar-r3 committed:
Rename FlowStateMachineReturnable to FlowStateMachineHandle

kyriathar-r3 committed:
Move logic from FlowStateMachine to FlowStateMachineHandle to get rid of downcasts and make it nullable; assert that logic is not null wherever it was used

kyriathar-r3 committed:
Move clientID to FlowStateMachineHandle and get rid of FlowStateMachineClientIdResult

kyriathar-r3 committed:
More replacing FlowStateMachine with FlowStateMachineHandle

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AppServiceHubImpl.kt:
```diff
@@ -78,7 +78,8 @@ internal class AppServiceHubImpl<T : SerializeAsToken>(private val serviceHub: S
                 id = stateMachine.id,
                 returnValue = stateMachine.resultFuture,
-                progress = stateMachine.logic.track()?.updates ?: Observable.empty()
+                progress = stateMachine.logic?.let { it.track()?.updates ?: Observable.empty() }
```
I think it could just be the empty `Observable`

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -237,27 +238,35 @@ internal class CordaRPCOpsImpl(
-                stepsTreeFeed = stateMachine.logic.trackStepsTree()
-        )
+                progress = logic.track()?.updates?.filter { !it.startsWith(STRUCTURAL_STEP_PREFIX) } ?: Observable.empty(),
+                stepsTreeIndexFeed = logic.trackStepsTreeIndex(),
```
I think rather than throwing an exception, we can fill in these fields.  The `steps...` fields are nullable and again we can use `empty()` for the `progress`.

kyriathar replied:
Done.

rick-r3 reviewed (commented):
A few changes to the null handling, but otherwise looks good.

kyriathar-r3 committed:
Address PR comments

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/FlowStateMachine.kt:
```diff
@@ -11,10 +11,19 @@ import net.corda.core.node.ServiceHub
 @DeleteForDJVM
 @DoNotImplement
-interface FlowStateMachine<FLOWRETURN> {
+interface FlowStateMachine<FLOWRETURN>: FlowStateMachineHandle<FLOWRETURN> {
```
Style: should be a space before the `:`.  When we split two types with a `:` it should be surrounded with whitespace.  When it is return type or `val` or `var` there's no space before the `:`.

kyriathar replied:
Thanks, done.

rick-r3 reviewed (commented):
LGTM.  One tiny formatting thing.

kyriathar-r3 committed:
First iteration of holding client ids to flow ids mapping into StateMachineManager and reclaiming them

kyriathar-r3 committed:
Remove unwanted comments

kyriathar-r3 committed:
Adding space

kyriathar-r3 committed:
Aling test names

kyriathar-r3 committed:
Add test asserting that only one flow starts upon concurrent requests with same client id

kyriathar-r3 committed:
Add hook needed by test in SingleThreadedStateMachineManager

kyriathar-r3 committed:
Use one semaphore among all threads

kyriathar-r3 committed:
Fix semantically erroneous code

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -22,6 +22,7 @@ import net.corda.nodeapi.internal.persistence.currentDBSession
 import org.apache.commons.lang3.ArrayUtils.EMPTY_BYTE_ARRAY
 import org.apache.commons.lang3.exception.ExceptionUtils
 import org.hibernate.annotations.Type
+import sun.plugin.dom.exception.InvalidStateException
```
Wrong one I reckon ?

kyriathar replied:
Replaced with correct one.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,14 +263,32 @@ class SingleThreadedStateMachineManager(
+        // therefore just return its future here and don't go any further down
+        return context.clientID?.let { clientID ->
+            mutex.locked {
+                clientIDsToFlowIds[clientID]?.let {
```
If think if you use `clientIDsToFlowIds[clientID].computeIfAbsent()` to put in place the flow ID and the Future in a way that would also work in a concurrent environment.   Then after `startFlowInternal` wire up the Future to the Future returned by that.

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -795,6 +822,10 @@ class SingleThreadedStateMachineManager(
                     incrementLiveFibers()
                     unfinishedFibers.countUp()
+                    // just point to new flow in clientIDsToFlowIds to allow previous one to get GCed
+                    flow.fiber.clientID?.let {
```
See my earlier comment about using `putIfAbsent` much earlier in the call chain.  The Future doesn't have to be the same future.... it could be one that is chained off this one.

kyriathar replied:
This bit of code is now completely removed since we now rely on `CordaFuture<FlowStateMachine<A>>`  returned from `startFlowInternal`.

kyriathar-r3 committed:
Adding client id entry early and then wiring its future further below; eliminates race condition that would apply on ENT: two concurrent requests could check for a client id at the same time, if that id was not there, they would start two different flows for the same client id

kyriathar-r3 committed:
Make clientIDsToFlowIds hold a future of FlowStateMachine instead to leverage the already created future created in startFlowInternal

kyriathar-r3 committed:
Remove unused import

kyriathar-r3 committed:
Remove uneeded locking

kyriathar-r3 committed:
Replace wrong exception type

kyriathar-r3 committed:
Add flow -started with client id- status

kyriathar-r3 committed:
Ignore test fow now; to be brought back upong implementing CORDA-3681

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +265,50 @@ class SingleThreadedStateMachineManager(
+                            }
+                        }
+                    return@startFlow future as CordaFuture<FlowStateMachine<A>>
+                } ?: clientIDsToFlowIds.putIfAbsent(clientID, FlowWithClientIdStatus.Active(openFuture()))
```
I think you should be able to wrap all of these lines into a single `clientIDstoFlowIds.compute(clientID) { }`.  This also has benefits when it comes to concurrency later.

rick-r3 replied:
Still needs that `compute` !

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +265,50 @@ class SingleThreadedStateMachineManager(
+            flowStart = FlowStart.Explicit,
+            ourIdentity = ourIdentity ?: ourFirstIdentity,
+            deduplicationHandler = deduplicationHandler
+        ).also {
```
I find myself wondering what happens if `startFlowInternal` throws an exception in some way?  Can it throw an exception ?  Do we need to worry ?  

If it did throw an exception, you'd never start this flow.  Resubmitting a start flow would de-duplicate and hand back the future from `clientIDsToFlowIds` but there's no code every going to _actually_ start the flow!

So if `startFlowInternal` can throw, we need to do something with the content of `clientIDsToFlowIds`.  Either complete the Future with an Exception, or remove the map entry, or something.

lankydan replied:
I don't think I have any testing around this atm, but there is nothing stopping `startFlowInternal` from throwing an exception, so we should take this into account. `onSessionInit` calls `startFlowInternal` and catches the exception. All that being said, I think the likely hood of it throwing is small.

lankydan replied:
I would say if we failed to even start the flow we should complete the future with an exception and remove the map entry.

lankydan replied:
Probably need a try catch around this and sort out the data structures in the catch

mnesbit replied:
My hope was that the pre-start checkpoint approach would write the flow details and flush the client id to DB with the meta. This would only throw if we couldn't deserialize the instructions and would always allow you to rebind a result future. Subsequent call to start the flow would then go down error/complete/still processing route.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineState.kt:
```diff
@@ -380,3 +383,10 @@ sealed class SubFlowVersion {
+sealed class FlowWithClientIdStatus {
+    data class Active(val flowStateMachineFuture: OpenFuture<FlowStateMachine<*>>) : FlowWithClientIdStatus()
+    data class Removed(val flowId: StateMachineRunId, val removedStatus: Status) : FlowWithClientIdStatus() {
+        enum class Status { SUCCEEDED, FAILED }
```
Does this need its own `Status` type or could it just use a boolean (`val succeeded: Boolean`) or the existing `FlowStatus` ?  Do we actually need to represent any status where there isn't a Fiber to link the `Future` to?  e.g. paused.

kyriathar replied:
Done.

lankydan commented on core/src/main/kotlin/net/corda/core/context/InvocationContext.kt:
```diff
@@ -105,14 +107,16 @@ data class InvocationContext(
         actor: Actor? = this.actor,
         externalTrace: Trace? = this.externalTrace,
-        impersonatedActor: Actor? = this.impersonatedActor
+        impersonatedActor: Actor? = this.impersonatedActor,
```
Not really to do with your change, but why do we even have this copy? For Java apis?

If we are concerned with people other than us (Corda) creating `InvocationContext`s then we need to be concerned about compatibility. I don't think we need to care about this though as realisically no developer should be manually creating an `InvocationContext`.

kyriathar replied:
I think that's there to restrict the caller from passing in `arguments` when copying.

lankydan commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -264,6 +264,9 @@ interface CordaRPCOps : RPCOps {
     fun <T> startFlowDynamic(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T>
 
+    @RPCReturnsObservables
+    fun <T> startFlowDynamicWithClientId(clientID: String, logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T>
```
Naming wise, do we need to say `WithClientId`? I get that it shows that a client id is being added, but in terms of overloads, we don't need the name as this function will still compile while being named `startFlowDynamic`.

I was originally thinking that the client id should go after the `logicType` but I guess without naming it `withClientId` there would then be no way to distinguish between this and the non client id version (due to the varags)

rick-r3 replied:
Not sure whether the RPC infra. supports overloads or not.  There's a method name that gets transported across the wire but not sure how much it deals with overloads.

Anyway, because the semantics of this are quite different I would prefer it wasn't an overload.  I think overloads suggest same behaviour, different args.  WDYT?

lankydan replied:
I still prefer it with an overloaded name, if it doesn't work then that is obviously a deal-breaker.

I think because of the order of the arguments, it distinguishes itself enough from the version without the id.

kyriathar replied:
RPC infra does not support overloading.

lankydan replied:
k, keep it as it is then

lankydan commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -264,6 +264,9 @@ interface CordaRPCOps : RPCOps {
 
+    @RPCReturnsObservables
+    fun <T> startFlowDynamicWithClientId(clientID: String, logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T>
+
```
All the other start flow functions at the end this file need changing, e.g, `CordaRPCOps.startFlow`, `CordaRPCOps.startTrackedFlow`.

rick-r3 replied:
I consciously didn't want to offer those methods in this form because they would be hard to implement.

kyriathar replied:
Unless i m missing something the below seems to be working fine: 

```kotlin 
@Suppress("unused")
inline fun <T, A, B, C, D, reified R : FlowLogic<T>> CordaRPCOps.startFlowDynamicWithClientId(
    clientId: String,
    @Suppress("unused_parameter")
    flowConstructor: (A, B, C, D) -> R,
    arg0: A,
    arg1: B,
    arg2: C,
    arg3: D
): FlowHandleWithClientId<T> = startFlowDynamicWithClientId(clientId, R::class.java, arg0, arg1, arg2, arg3)
```

kyriathar replied:
Done -added generic overloaded methods for type safe invocation of flows with client id from Kotlin-.

rick-r3 replied:
Ah, I wasn't talking about those (which we should have I think).  I thought @lankydan was talking about tracked versions of the start flow method etc.

lankydan replied:
👍 im fine with leaving out the tracked versions for now. Just wanted to make sure we can do things like `startFlow` from kotlin, like we currently do, but have a version that accepts client id as well.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -1268,13 +1269,26 @@ internal fun logVendorString(database: CordaPersistence, log: Logger) {
-class FlowStarterImpl(private val smm: StateMachineManager, private val flowLogicRefFactory: FlowLogicRefFactory) : FlowStarter {
-    override fun <T> startFlow(event: ExternalEvent.ExternalStartFlowEvent<T>): CordaFuture<FlowStateMachine<T>> {
-        smm.deliverExternalEvent(event)
+class FlowStarterImpl(private val smm: StateMachineManager, private val flowLogicRefFactory: FlowLogicRefFactory, private val maxClientIdLength: Int) : FlowStarter {
```
Can we format these class properties onto new lines.

kyriathar replied:
Done.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -1268,13 +1269,26 @@ internal fun logVendorString(database: CordaPersistence, log: Logger) {
+        // check clientID early, so that we don't get further down if it can't be saved in the database
+        val clientID = event.context.clientID
+        if (clientID != null && clientID.length > maxClientIdLength) {
+            event.wireUpFuture(doneFuture(object : FlowStateMachineHandle<T> {
```
Can we move the creation of the `doneFuture` into a separate function.

rick-r3 replied:
I think we shouldn't use `object : FlowStateMachineHandle<T>` here because in the real world this needs to be `@CordaSerializable` and available to the client.  So it needs to use a real class from `core` module I think.  Pretty sure you can use `FlowHandleWithClientIdImpl` ?

We need to test the API from a real RPC client I guess so that we don't miss this sort of thing.  There might be others.

kyriathar replied:
This bit of code is now changed (`doneFuture` is removed) as we are throwing a plain exception instead.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -1268,13 +1269,26 @@ internal fun logVendorString(database: CordaPersistence, log: Logger) {
-        smm.deliverExternalEvent(event)
+class FlowStarterImpl(private val smm: StateMachineManager, private val flowLogicRefFactory: FlowLogicRefFactory, private val maxClientIdLength: Int) : FlowStarter {
+    override fun <T> startFlow(event: ExternalEvent.ExternalStartFlowEvent<T>): CordaFuture<out FlowStateMachineHandle<T>> {
+        // check clientID early, so that we don't get further down if it can't be saved in the database
```
I think we can get away without the comment.

kyriathar replied:
Done -comment removed-.

williamvigorr3 commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -832,6 +837,147 @@ class FlowFrameworkTests {
+    // TODO: We set a new clientId (in StateMachineManager.addAndStartFlow) much later than we check if it exists; What will happen
+    //  if two concurrent requests come in with the same client id (these two actions need to be done in one place).
+    @Test
+    fun `no new flow starts if the client id provided pre exists`() {
```
Maybe this stuff should go in its own separate file?

lankydan replied:
Agreed, I dont want us to keep adding tests to this file when doing changes that achieve a single goal.

kyriathar replied:
Yes all these tests will be moved to a dedicated file.

kyriathar replied:
Done -client id tests were moved under newly introduced `FlowClientIdTests`-.

williamvigorr3 commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowMetadataRecordingTest.kt:
```diff
@@ -393,6 +398,19 @@ class FlowMetadataRecordingTest {
+            val nodeAHandle = startNode(providedName = ALICE_NAME, rpcUsers = listOf(user)).getOrThrow()
+            val rpc = CordaRPCClient(nodeAHandle.rpcAddress).start(user.username, user.password).proxy
+
+            assertFailsWith<CordaRuntimeException>("clientID cannot be longer than ${DBCheckpointStorage.MAX_CLIENT_ID_LENGTH} characters") {
```
Should we also have a "unit" test in DBCheckpointStorageTests for this or is that overkill?

lankydan replied:
I think its overkill. Plus we havent added the code to actually save it to the db just yet.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/AppServiceHubImpl.kt:
```diff
@@ -78,7 +78,7 @@ internal class AppServiceHubImpl<T : SerializeAsToken>(private val serviceHub: S
                 id = stateMachine.id,
                 returnValue = stateMachine.resultFuture,
-                progress = stateMachine.logic.track()?.updates ?: Observable.empty()
+                progress = stateMachine.logic?.let { it.track()?.updates } ?: Observable.empty()
```
We don't need this `let`

kyriathar replied:
Done -`let` removed-.

lankydan commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -237,27 +238,32 @@ internal class CordaRPCOpsImpl(
     }
 
-    private fun <T> startFlow(logicType: Class<out FlowLogic<T>>, args: Array<out Any?>): FlowStateMachine<T> {
+    override fun <T> startFlowDynamicWithClientId(clientID: String, logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T> {
```
I'll put it here since I just noticed, we need to be consistent with naming, I think it should be `clientId`. Whatever we choose, functions should follow the same naming, `withClientId` does not match `clientID` at the moment. This solves itself if we rename the variable of cource.

kyriathar replied:
Done -renamed to `clientId` in all places-.

lankydan commented on node/src/main/kotlin/net/corda/node/services/api/ServiceHubInternal.kt:
```diff
@@ -234,7 +234,7 @@ interface FlowStarter {
             logicType: Class<out FlowLogic<T>>,
             context: InvocationContext,
-            vararg args: Any?): CordaFuture<FlowStateMachine<T>>
+            vararg args: Any?): CordaFuture<out FlowStateMachineHandle<T>>
```
Fix this formatting if we are changing this.

kyriathar replied:
Done.

lankydan commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -264,6 +264,9 @@ interface CordaRPCOps : RPCOps {
     @RPCReturnsObservables
     fun <T> startFlowDynamic(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T>
 
+    @RPCReturnsObservables
```
Should we create a type alias for `clientId`? Rather than passing `String` everywhere. It could be `RPCClientId`, `ClientGeneratedId` or `UserSuppliedId`/`UserSuppliedIdentifier`?

rick-r3 replied:
Out of interest, does that present any issues with Java with this being an API?  I guess it just compiles down to `String`.  I'm not sure it's needed here because the field has a name.  Makes more sense when you have things like maps and you want to indicate what the key or the value means.

lankydan replied:
I agree, but i think it does make things a bit clearer. We do the same thing for `AttachmentId` being a type alias for `SecureHash`. Im not super fussed on this one.

kyriathar replied:
After offline discussion we are not going for the type alias after all (Java does not see it).

lankydan commented on node/src/main/kotlin/net/corda/node/services/logging/ContextualLoggingUtils.kt:
```diff
@@ -13,6 +13,12 @@ internal fun InvocationContext.pushToLoggingContext() {
     externalTrace?.pushToLoggingContext("external_")
     impersonatedActor?.pushToLoggingContext("impersonating_")
+
+    clientID?.let {
```
If we use a type alias, we can create an extension function to follow the existing convensions in this class.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -547,7 +547,11 @@ class DBCheckpointStorage(
     private fun InvocationContext.getFlowParameters(): List<Any?> {
         // Only RPC flows have parameters which are found in index 1
```
This comment needs updating.

kyriathar replied:
Done.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -547,7 +547,11 @@ class DBCheckpointStorage(
         // Only RPC flows have parameters which are found in index 1
         return if (arguments.isNotEmpty()) {
-            uncheckedCast<Any?, Array<Any?>>(arguments[1]).toList()
+            when {
```
Is it too smart to change it to `uncheckedCast<Any?, Array<Any?>>(arguments[arguments.size - 1]).toList()`?

Downside to that is loss of the exception if the size is different for some reason.

Is the code below better or worse? I'm not sure, your code is more explicit.

```kotlin
arguments.run {
  check(size == 2 || size == 3) { "Unexpected argument number provided in rpc call" }
  uncheckedCast<Any?, Array<Any?>>(this[size - 1]).toList()
}
```

lankydan replied:
Or even `uncheckedCast<Any?, Array<Any?>>(last()).toList()`

kyriathar replied:
Done -changed as proposed-.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -132,6 +132,8 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
     override val ourIdentity: Party get() = transientState!!.value.checkpoint.checkpointState.ourIdentity
     override val isKilled: Boolean get() = transientState!!.value.isKilled
+    override val clientID: String?
+        get() = transientState?.value?.checkpoint!!.checkpointState.invocationContext.clientID
```
Lets stick with the convention of the other propterties and do `transientState!!`

kyriathar replied:
Done.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +265,50 @@ class SingleThreadedStateMachineManager(
             deduplicationHandler: DeduplicationHandler?
-    ): CordaFuture<FlowStateMachine<A>> {
+    ): CordaFuture<out FlowStateMachineHandle<A>> {
+        beforeClientIDCheck?.invoke()
```
Are we keeping this hook? Need to think about whether we want to keep adding hooks like this to enable testing or to use byteman to achieve similar behaviour.

kyriathar replied:
I only used these hooks to write a test that reproduces concurrent requests issue and then build the actual code to solve it. Also removed locking on a mutex so i can actually reproduce that 100 %. After adding the code that fixed that issue, i put back the locking on that mutex and test still works correctly (only not truly concurrent anymore). Therefore, i would say, these should be removed... i guess.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +265,50 @@ class SingleThreadedStateMachineManager(
+                        when (flowWithClientIdStatus) {
+                            is FlowWithClientIdStatus.Active -> flowWithClientIdStatus.flowStateMachineFuture
+                            is FlowWithClientIdStatus.Removed -> {
+                                doneFuture(object : FlowStateMachineHandle<A> {
```
Can we extract this done future to a separate function.

kyriathar replied:
Done -new method was created and an additional call site for this method in `StateMachineManager.start`-.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -871,6 +917,11 @@ class SingleThreadedStateMachineManager(
         require(flow.fiber.id !in sessionToFlow.values) { "Flow fibre must not be needed by an existing session" }
         flow.resultFuture.set(removalReason.flowReturnValue)
+        flow.fiber.clientID?.let {
+            val oldClientIdFlowStatus = clientIDsToFlowIds[it]
```
We can extract this into a function as it is duplicated with the code below. Pass in the final status into the new function then reuse it.

kyriathar replied:
Done.

kyriathar-r3 committed:
On node restart flows -to run- with client id are hook-able

kyriathar-r3 committed:
Remove badly used generics

kyriathar-r3 committed:
Assert that only one flow is spawned upon multiple requests with same client id

kyriathar-r3 committed:
On node restart -paused- flows with client id are hook-able; NonResidentFlow now holds a resultFuture needed for FlowWithClientIdStatus.Active placeholder

kyriathar-r3 committed:
Remove trivial comment

kyriathar-r3 committed:
Remove uneeded comment

kyriathar-r3 committed:
Introduce remove client id API through which an RPC client can remove the client id to flow's result/ exception mapping

kyriathar-r3 committed:
Add FlowHandleWithClientId to return client id as well for correlation on client side

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -278,6 +281,13 @@ interface CordaRPCOps : RPCOps {
+     *
+     * @return whether the mapping was removed.
+     */
+    fun removeClientId(clientID: String): Boolean
```
Originally I was thinking this would be a method on `FlowHandleWithClientId` but the way you have it has advantages.  It would be a bit tedious for a clean up job to have to fetch the result, only to call remove.  So remove as you have it is actually better, yes.

I think it might be worth sketching out example code of how this might be used together with `startFlowDynamicWithClientId ` in a sort of reliable and clean exemplar client. e.g. should it be:

```
var success = false
while(!success) {
  try { 
    val handle = startFlowDynamicWithClientId()
    useResult(handle.clientID, handle.returnValue.get())
    removeClientId(handle.clientID) 
    success = true
  } catch {
    // Log or something about trying again
  }
}
```

I think there is an argument for saying that it should be obvious how to use this API and if we have to give some clear indicators in the documentation of `startFlowDynamicWithClientId` then we should.

kyriathar replied:
Recorded under https://r3-cev.atlassian.net/browse/ENT-5464 to be added in documentation.

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -278,6 +281,13 @@ interface CordaRPCOps : RPCOps {
+     * @return whether the mapping was removed.
+     */
+    fun removeClientId(clientID: String): Boolean
+
```
I think we could also offer a `reattachClientId` but you might just want to file that as a JIRA for the future ?

That method would fail if the client ID did not already exist, but would return the handle if it did.

kyriathar replied:
This is recorded in https://r3-cev.atlassian.net/browse/CORDA-3822. Adding your suggestion as a comment in that ticket.

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -278,6 +281,13 @@ interface CordaRPCOps : RPCOps {
     fun killFlow(id: StateMachineRunId): Boolean
 
+    /**
+     * Removes a flow's [clientID] to result/ exception mapping.
```
I think this should somehow reference `startFlowDynamicWithClientId` for "further information".

kyriathar replied:
Done.

rick-r3 commented on core/src/main/kotlin/net/corda/core/messaging/FlowHandle.kt:
```diff
@@ -28,6 +28,14 @@ interface FlowHandle<A> : AutoCloseable {
+    /**
+     * The [clientID] with which the client has started the flow.
+     */
+    val clientID: String
```
Should we also have a `remove` method here that just calls `removeClientId` with this `clientID` ?

kyriathar replied:
Recorded under https://r3-cev.atlassian.net/browse/CORDA-3889 for vNext.

rick-r3 replied:
As discussed, actually harder than it sounds in terms of RPC connection loss and testing.  So deferring such a method for another time.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -913,4 +992,10 @@ class SingleThreadedStateMachineManager(
+
+    override fun removeClientId(clientID: String): Boolean {
+        return mutex.locked {
+            clientIDsToFlowIds.remove(clientID) != null
```
Surely if the flow is running, we cannot remove ?  It would be confusing and what happens when the flow ends?

I think the flow needs to have ended before it is okay to remove.

kyriathar replied:
I thought of it as well, i think: If a flow is still running, then that request that initiated that flow or any other that comes in before the "remove" request will get back their `FlowHandleWithClientId` and these objects will hold a reference to the `resultFuture`. So once we request to remove the client id we just remove the placeholder while the RPC clients still holding their `FlowHandleWithClientId`, unless i m missing something.

rick-r3 replied:
But the idea of the remove is to clean up the result / exception in the database and the main checkpoint record including the meta data, not the in-memory map (although you need that too).  You can't do that until the flow finishes.

rick-r3 replied:
i.e. I think you can only remove if it is `FlowWithClientIdStatus.Removed`

kyriathar replied:
Done.

kyriathar-r3 committed:
Use Map.compute to create flowStateMachineFuture or insert FlowWithClientIdStatus.Active to clientIDsToFlowIds

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -292,26 +292,27 @@ class SingleThreadedStateMachineManager(
-                    val future =
-                        when (flowWithClientIdStatus) {
-                            is FlowWithClientIdStatus.Active -> flowWithClientIdStatus.flowStateMachineFuture
+                clientIDsToFlowIds.compute(clientID) { _, flowWithClientIdStatus ->
```
I think I would rename `future` to `existingFuture` and `flowWithClientIdStatus` to `existingStatus`

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -292,26 +292,27 @@ class SingleThreadedStateMachineManager(
-                        when (flowWithClientIdStatus) {
-                            is FlowWithClientIdStatus.Active -> flowWithClientIdStatus.flowStateMachineFuture
+                clientIDsToFlowIds.compute(clientID) { _, flowWithClientIdStatus ->
+                    flowWithClientIdStatus?.let {
```
I think a straightforward `if(existingStatus != null) { } else { }` is easier to read.  Remember the if/else has a return value, which here we can use as the result of the compute.

kyriathar replied:
Done.

rick-r3 reviewed (commented):
Made some suggestions, but yes, this is what I was thinking so that it is all atomic inside `compute`

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -188,6 +200,27 @@ class SingleThreadedStateMachineManager(
             }
         }
+
+        // at the moment we have RUNNABLE, HOSPITALIZED and PAUSED -> TODO: RESULTED AND FAILED still need to be fetched
```
What about incompatible flows?

kyriathar replied:
As discussed offline this stays as TODO for now, documented under https://r3-cev.atlassian.net/browse/CORDA-3897.

kyriathar-r3 committed:
Rename vars

kyriathar-r3 committed:
Replace let with if else statement

kyriathar-r3 committed:
Allow client id mapping removal only as soon as the flow gets removed

kyriathar-r3 committed:
Update kdocs

kyriathar-r3 committed:
Throw plain exception on -bigger than max client id length- client id

kyriathar-r3 committed:
Remove uneeded comment

kyriathar-r3 committed:
Code formatting

kyriathar-r3 committed:
Sort client id mappings before setting the result futures

kyriathar-r3 committed:
Update test to assert -running- flows are re-hookable when called with the same client id

kyriathar-r3 committed:
Make test more precise in terms of when threads actually execute things

kyriathar-r3 committed:
Fix test; clear hook that was causing test thread to hang in other tests

kyriathar-r3 committed:
Remove not needed let method

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +287,54 @@ class SingleThreadedStateMachineManager(
+        ).also {
+            if (clientID != null) {
+                // wire up this future to the clientIDsToFlowIds[clientID] future
+                val active = mutex.content.clientIDsToFlowIds[clientID] as? FlowWithClientIdStatus.Active
```
If you move `existingFuture` outside of `if (clientID != null) {` further up in this method (line 295-ish), you don't need to retrieve it again here.  Also, you are retrieving without taking the mutex which is questionable, and using the one retrieved further up under the mutex also fixes that.

kyriathar replied:
Good catch, i must had thought of it like at this point only one thread can be operating on this block of code since all others will find the respective placeholder get that and return (in concurrent environment), however, if a  flow starts with this `clientID` and in theory it could finish at the same time and update the same entry in the `Map`, as this thread accesses that entry from the `Map`.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -870,6 +942,11 @@ class SingleThreadedStateMachineManager(
+        flow.fiber.clientID?.let {
+            val oldClientIdFlowStatus = clientIDsToFlowIds[it]
+            require (oldClientIdFlowStatus != null && oldClientIdFlowStatus is FlowWithClientIdStatus.Active)
+            clientIDsToFlowIds[it] = FlowWithClientIdStatus.Removed(flow.fiber.id, FlowWithClientIdStatus.Removed.Status.SUCCEEDED)
```
This and the retrieval 2 lines above need to have the mutex when they do this.  Maybe this already has the mutex (I didn't check) via the call site.  Also try and use `compute` to make it enterprise-proof.

kyriathar replied:
It did hold the mutex from call site which is `removeFlow`. However, indeed it needs to be ent-proof thefore changing it to `compute`, good catch thanks.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -881,6 +958,11 @@ class SingleThreadedStateMachineManager(
+        flow.fiber.clientID?.let {
+            val oldClientIdFlowStatus = clientIDsToFlowIds[it]
+            require (oldClientIdFlowStatus != null && oldClientIdFlowStatus is FlowWithClientIdStatus.Active)
+            clientIDsToFlowIds[it] = FlowWithClientIdStatus.Removed(flow.fiber.id, FlowWithClientIdStatus.Removed.Status.FAILED)
```
This and the retrieval 2 lines above need to have the mutex when they do this. Maybe this already has the mutex (I didn't check) via the call site. Also try and use compute to make it enterprise-proof.

kyriathar replied:
Same as above.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -913,4 +995,15 @@ class SingleThreadedStateMachineManager(
+
+    override fun removeClientId(clientID: String): Boolean {
+        return mutex.locked {
+            val existingStatus = clientIDsToFlowIds[clientID]
```
Are you able to use `compute` again here to do this atomically?

kyriathar replied:
Done -left `mutex.locked` in, so it can't happen whilst a flow with the same client id is in the process of `removeFlow `-.

kyriathar-r3 committed:
Replace transientState? with transientState!! to be aligned with rest of accesses

kyriathar-r3 committed:
Update comment

kyriathar-r3 committed:
Use local variable to hold newly put future in clientIDsToFlowIds; solves potential race issue in which a flow could finish fast and replace the same entry in clientIDsToFlowIds with 'Removed' before the current thread returns in startFlow

kyriathar-r3 committed:
Set 'FlowWithClientIdStatus' to 'Removed' (SUCCEEDED or FAILED) inside compute

kyriathar-r3 committed:
Use compute in 'removeClientId' to make it atomic for concurrent environment

kyriathar-r3 committed:
Put duplicate code into method

kyriathar-r3 committed:
Rename 'clientID' to 'clientId' in all places

kyriathar-r3 committed:
Format code

kyriathar-r3 committed:
Update TODO comment

kyriathar-r3 committed:
Extract common code into new method

kyriathar-r3 committed:
Replace Removed.Status with Boolean as we will only have statuses 'succeeded' and 'failed'

kyriathar-r3 committed:
Add generic overloaded methods for type safe invocation of flows with client id from Kotlin

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -254,15 +283,46 @@ class SingleThreadedStateMachineManager(
+            ourIdentity = ourIdentity ?: ourFirstIdentity,
+            deduplicationHandler = deduplicationHandler
+        ).also {
+            if (clientId != null) {
```
I think change this to be `if(newFuture != null) {` and then you don't need the null check in the line below etc.

Also, set the type of `newFuture` to be the one you want/cast to here.  e.g. Further up, rather than

```
val newFuture: CordaFuture<out FlowStateMachineHandle<out Any?>>? = null
....
FlowWithClientIdStatus.Active(openFuture()).also { newFuture = it.flowStateMachineFuture }
```

have 

```
val newFuture: OpenFuture<out FlowStateMachineHandle<out Any?>>? = null
...
newFuture = openFuture()
FlowWithClientIdStatus.Active(newFuture)
```

kyriathar replied:
Done.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -913,4 +975,55 @@ class SingleThreadedStateMachineManager(
+        )
+
+    override fun removeClientId(clientId: String): Boolean {
+        var removed = false
```
Rather than have remove flag, just `return clientIdsToFlowIds.compute(clientId) { ... } == null` ?  i.e. if it is removed, the return value of `compute` is null.

rick-r3 replied:
Keep the mutex through, but that returns the value of the thing inside it.

kyriathar replied:
The thing is, what if there is currently no value in the Map for the given `clientId` . `compute` will still return null but nothing actually removed.

kyriathar-r3 committed:
Minor code refactoring

mnesbit commented on core/src/main/kotlin/net/corda/core/internal/FlowStateMachine.kt:
```diff
@@ -11,10 +11,19 @@ import net.corda.core.node.ServiceHub
+@DeleteForDJVM
+@DoNotImplement
+interface FlowStateMachineHandle<FLOWRETURN> {
+    val logic: FlowLogic<FLOWRETURN>?
```
Out state machine observables are one of the things that prevents us doing horizontal scaling of the node, so keen not to use those features if we can avoid.

kyriathar-r3 committed:
Minor refactoring to 'newFuture' setting

kyriathar-r3 committed:
Check for 'newFuture' being null instead of 'clientId'

kyriathar-r3 committed:
Remove implemented TODO

kyriathar-r3 committed:
Move client id tests under dedicated testing class

kyriathar-r3 committed:
Update test; better sync threads so that second request waits until flow has retried

kyriathar-r3 committed:
Adding 'this' keyword for consistency

kyriathar-r3 committed:
Move 'clientId' to be the last property of InvocationContext

kyriathar-r3 committed:
Revert "Adding 'this' keyword for consistency"

This reverts commit 549b7686617df4cc1900cd33ad759b74da8a5998.

kyriathar-r3 committed:
Remove 'clientId' from InvocationContext overloaded 'copy' parameters

kyriathar-r3 committed:
Made 'InvocationContext.arguments' nullable so that - >= 4.6 version - RPC client are compatible against - < 4.6 version - nodes

kyriathar-r3 committed:
If 'startFlowInternal' throws remove client id entry from client id map; Prevents subsequent requests from de-duplicating and getting a reference to a -never started flow- future

kyriathar-r3 committed:
If 'startFlowInternal' throws complete 'FlowStateMachineHandle' future

kyriathar-r3 committed:
Align code after rebase

kyriathar-r3 committed:
Move the removal and completion of 'startedFuture' before setting of client id as failed

kyriathar-r3 committed:
Minor update to test name - comment

kyriathar-r3 committed:
Remove file - leftover from rebase

kyriathar-r3 committed:
Remove wrong comment

kyriathar-r3 committed:
Minor reordering to test code

kyriathar-r3 committed:
Test vars renamings

kyriathar-r3 committed:
Minor modifications on test names

kyriathar-r3 committed:
Minor comment update

kyriathar-r3 committed:
Add timeouts to tests

kyriathar-r3 committed:
Remove unused imports and line break

kyriathar-r3 committed:
Fix detekt issues

kyriathar-r3 committed:
Make detekt happy

kyriathar-r3 committed:
Update detekt baseline

kyriathar-r3 committed:
Fix compilation error; Replace 'FlowStateMachine' with 'FlowStateMachineHandle'

kyriathar-r3 committed:
Remove unneeded comment

kyriathar commented:
run tests: slow

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -229,21 +262,60 @@ internal class SingleThreadedStateMachineManager(
+                clientIdsToFlowIds.remove(clientId)
+                newFuture?.setException(t)
+            }
+            throw t
```
I wonder if we should not do a `throw` here for when there's a `clientId`?  Here's my thinking:  With the throw there, the behaviour differs between the one who inserted into `clientIdsToFlowIds` vs. someone who encounters an entry already in the map.  i.e. if a caller just picks up the future from the map, they'll see the exception only via the future. Whereas in this instance if you are also the creator of the future, you see it direct, and via the future.

I would have thought the simplicity of a single outcome would be better.  Maybe only throw if `newFuture` is null ?  e.g. `newFuture?.setException(t) ?: throw t`.

I think keeping the throw as a fallback when not called with `clientId` is important, as we don't want to change the behaviour for other code paths (but we know we'll do a `getAndThrow` for the client ID path, so can channel it into the future without worry).

kyriathar replied:
Tested this and it has the exact same behavior on the client side in either case. As discussed offline, leaving it as is because its a bit more readable, but will add a comment explaining why it has the same behavior.

kyriathar-r3 committed:
Add explanatory comment for throwing plain exception in 'startFlow'

kyriathar-r3 committed:
Remove uneeded covariants

kyriathar commented:
run tests: slow

lankydan commented:
Before I forget, do we need to add these new functions to the `current-api.txt` file thingy?

rick-r3 commented:
@lankydan Not until the next release (although apparently we didn't update after 4.5, but think @fowlerrr is looking into that - maybe).

rick-r3 commented:
I think code owners implies @blsemo has to approve the change to `DBCheckpointStorage` (from what I can tell).  Is that for schema related things?  @blsemo can you approve it please?

josecoll commented:
Please please please remove me from this distribution list.

Jose Coll | R3. | Senior Software Developer
2 London Wall Place | 12th Floor | London | EC2Y 5AU
O: +44 207 113 1460
jose.coll@r3.com<mailto:jose.coll@r3.com> . www.r3.com<http://www.r3.com/>

[/var/folders/48/bn9ym1093tdbfw4zp2wzf4kc0000gn/T/com.microsoft.Outlook/WebArchiveCopyPasteTempFiles/cidimage014.jpg@01D59639.1D973A10]
Unlock new value and transform your business.
Find out how here<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.r3.com%2F&data=02%7C01%7C%7Cae83596c36974bfdcd4508d7647a2cf1%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637088348329481929&sdata=BzTWSc79ZsjBrfWX3QXziRznhQfJYqbBlahjdbpnsqk%3D&reserved=0>.

From: Rick Parker <notifications@github.com>
Reply-To: corda/corda <reply@reply.github.com>
Date: Thursday, 16 July 2020 at 10:09
To: corda/corda <corda@noreply.github.com>
Cc: Jose Coll <jose.coll@r3.com>, Review requested <review_requested@noreply.github.com>
Subject: Re: [corda/corda] CORDA-3809 Expose client side unique RPC ID for flow starts (#6307)


I think code owners implies @blsemo<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fblsemo&data=02%7C01%7C%7C1945dc257a104224971508d82967ed4c%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637304873629517768&sdata=ecCo%2FsXRETkeohS%2Bc7pJ48iiRy4gIfO%2FFfi0G%2FyzWSc%3D&reserved=0> has to approve the change to DBCheckpointStorage (from what I can tell). Is that for schema related things? @blsemo<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fblsemo&data=02%7C01%7C%7C1945dc257a104224971508d82967ed4c%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637304873629517768&sdata=ecCo%2FsXRETkeohS%2Bc7pJ48iiRy4gIfO%2FFfi0G%2FyzWSc%3D&reserved=0> can you approve it please?

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcorda%2Fcorda%2Fpull%2F6307%23issuecomment-659275206&data=02%7C01%7C%7C1945dc257a104224971508d82967ed4c%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637304873629527756&sdata=dAwHGbm3pz5bbd9%2FDHb6eLlEo963bBf6%2B0xIzSkdNLE%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFTXT5TAMXY7H65L72CYDDDR327UBANCNFSM4NRRRHHA&data=02%7C01%7C%7C1945dc257a104224971508d82967ed4c%7Ca4be1f2e2d10419587cd736aca9b672c%7C0%7C0%7C637304873629527756&sdata=hnR3Za3nT%2BD2NA%2FSyVcW%2F5kLqPKWQLgswx7wZSL40Dw%3D&reserved=0>.
